### PR TITLE
IGNITE-14853

### PR DESCRIPTION
**Pull request by ascherbakoff**

### CODE REVIEW

File: modules/raft/src/test/java/org/apache/ignite/raft/jraft/core/TestCluster.java [hunk 1]
```diff
@@ -47,17 +47,25 @@
...
-    private static final int ELECTION_TIMEOUT = 300;
+    /**
+     * Default election timeout.
+     * Important: due to sync disk ops (writing raft meta) during probe request processing this timeout should be high
+     * enough to avoid test flakiness.
+     */
+    private static final int ELECTION_TIMEOUT = 600;
 
-    private static final IgniteLogger LOG = IgniteLogger.forClass(TestCluster.class);
+    private static final Logger LOG = LoggerFactory.getLogger(TestCluster.class);
```
Comment (sashapolo) [inline modules/raft/src/test/java/org/apache/ignite/raft/jraft/core/TestCluster.java:68]: why did you remove the `IgniteLogger`?

File: modules/raft/src/test/java/org/apache/ignite/raft/jraft/core/TestCluster.java [hunk 2]
```diff
@@ -47,17 +47,25 @@
...
  */
 public class TestCluster {
-    /** Default election timeout. */
-    private static final int ELECTION_TIMEOUT = 300;
+    /**
+     * Default election timeout.
+     * Important: due to sync disk ops (writing raft meta) during probe request processing this timeout should be high
+     * enough to avoid test flakiness.
+     */
+    private static final int ELECTION_TIMEOUT = 600;
```
Comment (sashapolo) [inline modules/raft/src/test/java/org/apache/ignite/raft/jraft/core/TestCluster.java:66]: I think it is always nice to use the time units as part of the variable name, for example: `ELECTION_TIMEOUT_SECONDS`

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java [hunk 3]
```diff
@@ -46,17 +38,50 @@ public static void throwException(final Throwable t) {
...
+     * Note that this method follows includes {@link Throwable#getSuppressed()}
+     * into check.
+     *
+     * @param t Throwable to check (if {@code null}, {@code false} is returned).
+     * @param msg Message text that should be in cause.
+     * @param cls Cause classes to check (if {@code null} or empty, {@code false} is returned).
+     * @return {@code True} if one of the causing exception is an instance of passed in classes,
+     *      {@code false} otherwise.
+     */
+    @SafeVarargs
```
Comment (sashapolo) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java:54]: Looks like `@SafeVarargs` annotation is not needed here

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java [hunk 4]
```diff
@@ -46,17 +38,50 @@ public static void throwException(final Throwable t) {
...
+    public static boolean hasCause(@Nullable Throwable t, @Nullable String msg, @Nullable Class<?>... cls) {
+        if (t == null || cls == null || cls.length == 0)
+            return false;
+
+        assert cls != null;
```
Comment (sashapolo) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java:59]: this assert is redundant

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java [hunk 5]
```diff
@@ -46,17 +38,50 @@ public static void throwException(final Throwable t) {
...
+     * into check.
+     *
+     * @param t Throwable to check (if {@code null}, {@code false} is returned).
+     * @param msg Message text that should be in cause.
+     * @param cls Cause classes to check (if {@code null} or empty, {@code false} is returned).
+     * @return {@code True} if one of the causing exception is an instance of passed in classes,
+     *      {@code false} otherwise.
+     */
+    @SafeVarargs
+    public static boolean hasCause(@Nullable Throwable t, @Nullable String msg, @Nullable Class<?>... cls) {
```
Comment (sashapolo) [inline modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java:55]: `@Nullable Class<?>... cls` - this notation actually means "an array of Class instances that can *contain* nulls", which doesn't make a lot of sense in this method

File: modules/raft/src/test/java/org/apache/ignite/raft/jraft/test/TestUtils.java [hunk 6]
```diff
@@ -90,29 +88,11 @@ public static LogEntry mockEntry(final int index, final int term, final int data
...
-                while (addresses.hasMoreElements()) {
-                    InetAddress addr = addresses.nextElement();
-                    if (addr instanceof Inet4Address) {
-                        ip = addr.getHostAddress();
-                        break;
-                    }
-                }
-            }
-            return ip;
+            return Inet4Address.getLocalHost().getHostAddress();
```
Comment (sashapolo) [inline modules/raft/src/test/java/org/apache/ignite/raft/jraft/test/TestUtils.java:92]: This diff looks weird, shoudn't this code already be in `main`?

File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java (hunk 5 above)
Comment (ascherbakoff) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java:55]: The annotation is used to indicate the `null `is a valid value for the cls array parameter, so it makes sense to me.

Comment (sashapolo) [reply modules/raft/src/main/java/org/apache/ignite/raft/jraft/util/internal/ThrowUtil.java:54]: If you want to indicate that `null` is a valid value for the whole array, you should write `Class<?> @Nullable... cls` instead

**Review by sashapolo — APPROVED** [review]
